### PR TITLE
#100: implement repository commandlet

### LINKED ISSUE #100: Implement Commandlet for projects/repositories
As a IDEasy user, I want to be able to add projects/repositories so that I can get them setup automatically.
In `devonfw-ide` there is `devon project ...` and the setup automatically adds all active projects:
* https://github.com/devonfw/ide/blob/master/documentation/project.asciidoc
* https://github.com/devonfw/ide/blob/master/scripts/src/main/resources/scripts/command/project

This functionality is missing in IDEasy and shall be implemented with this story.

### PR DESCRIPTION
**Pull request by salimbouch**
closes #100 

- implements repository commandlet (old project commandlet)
- solves bug, findBinary not finding the binary that was just installed (see setPath in SystemPath)

### CODE REVIEW

File: cli/src/main/java/com/devonfw/tools/ide/commandlet/RepositoryCommandlet.java
Comment (hohwille) [inline cli/src/main/java/com/devonfw/tools/ide/commandlet/RepositoryCommandlet.java:88]:
```suggestion
    this.context.info("Importing repository from {} ...", repositoryFile.getFileName().toString());
```

File: cli/src/main/java/com/devonfw/tools/ide/commandlet/RepositoryCommandlet.java
Comment (hohwille) [inline cli/src/main/java/com/devonfw/tools/ide/commandlet/RepositoryCommandlet.java:35]:
```suggestion
    addKeyword("setup");
```

File: cli/src/main/java/com/devonfw/tools/ide/io/FileAccess.java
Comment (hohwille) [inline cli/src/main/java/com/devonfw/tools/ide/io/FileAccess.java:139]:
```suggestion
   * @return a list of paths that satisfy the provided {@link Predicate}. Will be {@link List#isEmpty() empty} if no match was found but is never {@code null}.
```

File: cli/src/main/resources/nls/Ide_de.properties
Comment (hohwille) [inline cli/src/main/resources/nls/Ide_de.properties:8]:
```suggestion
cmd-repository=Richtet das vorkonfigurierte Git Repository ein.
```

File: cli/src/main/resources/nls/Ide_de.properties
Comment (hohwille) [inline cli/src/main/resources/nls/Ide_de.properties:29]:
```suggestion
val-repository-repository=Der Name der Properties-Datei des vorkonfigurierten Git Repositories zum Einrichten. Falls nicht angegeben, werden alle aktiven Projekte eingerichtet. 
```

File: cli/src/main/resources/nls/Ide.properties
Comment (hohwille) [inline cli/src/main/resources/nls/Ide.properties:8]:
```suggestion
cmd-repository=setup the pre-configured git repository.
```

File: cli/src/main/resources/nls/Ide.properties
Comment (hohwille) [inline cli/src/main/resources/nls/Ide.properties:29]:
```suggestion
val-repository-repository=The name of the properties file of the pre-configured git repository to setup, omit to setup all active repositories.
```

**Review by hohwille — CHANGES_REQUESTED** [review]
@salimbouch thanks for your PR. You integrated the feature properly in the right code places and your approach looks fine 👍 
I added several review comments for improvement. Please have a look.

File: cli/src/main/java/com/devonfw/tools/ide/io/FileAccess.java
Comment (hohwille) [inline cli/src/main/java/com/devonfw/tools/ide/io/FileAccess.java:141]: actually this method does not only return files but also folders. It should therefore be named `getChildren` or `getChildrenOfDir`.

File: cli/src/main/java/com/devonfw/tools/ide/commandlet/RepositoryCommandlet.java
Comment (salimbouch) [reply cli/src/main/java/com/devonfw/tools/ide/commandlet/RepositoryCommandlet.java]: done

File: cli/src/main/java/com/devonfw/tools/ide/commandlet/RepositoryCommandlet.java
Comment (salimbouch) [reply cli/src/main/java/com/devonfw/tools/ide/commandlet/RepositoryCommandlet.java]: done

File: cli/src/main/java/com/devonfw/tools/ide/commandlet/RepositoryCommandlet.java
Comment (salimbouch) [reply cli/src/main/java/com/devonfw/tools/ide/commandlet/RepositoryCommandlet.java]: done

File: cli/src/main/java/com/devonfw/tools/ide/commandlet/RepositoryCommandlet.java
Comment (salimbouch) [reply cli/src/main/java/com/devonfw/tools/ide/commandlet/RepositoryCommandlet.java]: done :) this inconsistency happened because the method used to return null when no children were found and I forgot to change this accordingly 😅

File: cli/src/main/java/com/devonfw/tools/ide/commandlet/RepositoryConfig.java
Comment (salimbouch) [reply cli/src/main/java/com/devonfw/tools/ide/commandlet/RepositoryConfig.java:24]: done

File: cli/src/main/java/com/devonfw/tools/ide/common/SystemPath.java
Comment (salimbouch) [reply cli/src/main/java/com/devonfw/tools/ide/common/SystemPath.java:169]: No this is the PR where I first stumbled upon this bug.

File: cli/src/main/java/com/devonfw/tools/ide/io/FileAccess.java
Comment (salimbouch) [reply cli/src/main/java/com/devonfw/tools/ide/io/FileAccess.java]: done

File: cli/src/main/java/com/devonfw/tools/ide/commandlet/RepositoryCommandlet.java
Comment (salimbouch) [reply cli/src/main/java/com/devonfw/tools/ide/commandlet/RepositoryCommandlet.java]: done

File: cli/src/main/java/com/devonfw/tools/ide/commandlet/RepositoryCommandlet.java
Comment (salimbouch) [reply cli/src/main/java/com/devonfw/tools/ide/commandlet/RepositoryCommandlet.java]: I implemented a RepositoryProperty but not sure if that's the right way to do it. How can I access the context from a property? For now I gave the context as a parameter to the property which doesn't seem clean, plus should the resolving of the `Path` from the string value be done in a new method or from `getValue()?`

File: cli/src/main/java/com/devonfw/tools/ide/commandlet/RepositoryCommandlet.java
Comment (salimbouch) [reply cli/src/main/java/com/devonfw/tools/ide/commandlet/RepositoryCommandlet.java]: done

File: cli/src/main/java/com/devonfw/tools/ide/commandlet/RepositoryCommandlet.java
Comment (hohwille) [inline cli/src/main/java/com/devonfw/tools/ide/commandlet/RepositoryCommandlet.java:102]: In line #113 we are calling `gitPullOrClone`. It would be better to keep this closer together as other invocations like `mkdirs` may also log on debug meanwhile.
However, since this debug message is entirely generic, why not directly having it inside `gitPullOrClone` method implementation?

File: cli/src/main/java/com/devonfw/tools/ide/commandlet/RepositoryConfig.java
Comment (hohwille) [inline cli/src/main/java/com/devonfw/tools/ide/commandlet/RepositoryConfig.java:48]: fine but be aware that `Boolean.parseBoolean("true ")` will return `false` what might be surprising and can be hard to trace down.

File: cli/src/main/java/com/devonfw/tools/ide/commandlet/RepositoryCommandlet.java
Comment (hohwille) [inline cli/src/main/java/com/devonfw/tools/ide/commandlet/RepositoryCommandlet.java:132]: mhm. Actual `TODO` tags should be avoided on `main`.
I fully agree that we should implement the actual project import in a separate PR later.
According to clean code it would be better to remove this if/for block and instead create a new story issue for the use-case of importing projects.

File: cli/src/main/java/com/devonfw/tools/ide/commandlet/RepositoryCommandlet.java
Comment (hohwille) [inline cli/src/main/java/com/devonfw/tools/ide/commandlet/RepositoryCommandlet.java:130]: Is this `.project` file the one from the `Eclipse` import? I guess so. Using constants in `IdeContext` that have some JavaDoc is always better to make the code understandable...

**Review by hohwille — CHANGES_REQUESTED** [review]
@salimbouch Thanks for your PR and the great work 👍 
Looks good to me and you did nicely implement the existing behavior from the bash script and the new functionality. And thanks for fixing other issues like JavaDoc arg name, etc.

Surely I left several review comments for improvement.
Once resolved, you can reassign to me and I can approve and merge.
If you have questions or want to discuss something about the design feel free to get in touch.

File: cli/src/main/java/com/devonfw/tools/ide/commandlet/RepositoryCommandlet.java
Comment (salimbouch) [reply cli/src/main/java/com/devonfw/tools/ide/commandlet/RepositoryCommandlet.java]: done

File: cli/src/main/java/com/devonfw/tools/ide/commandlet/RepositoryCommandlet.java
Comment (salimbouch) [reply cli/src/main/java/com/devonfw/tools/ide/commandlet/RepositoryCommandlet.java]: done

File: cli/src/main/java/com/devonfw/tools/ide/property/RepositoryProperty.java
Comment (salimbouch) [reply cli/src/main/java/com/devonfw/tools/ide/property/RepositoryProperty.java]: done

File: cli/src/main/java/com/devonfw/tools/ide/property/RepositoryProperty.java
Comment (salimbouch) [reply cli/src/main/java/com/devonfw/tools/ide/property/RepositoryProperty.java:69]: done, implemented method findExistingFile in FileAccess.

File: cli/src/main/java/com/devonfw/tools/ide/commandlet/RepositoryCommandlet.java
Comment (salimbouch) [reply cli/src/main/java/com/devonfw/tools/ide/commandlet/RepositoryCommandlet.java:102]: done :)

File: cli/src/main/java/com/devonfw/tools/ide/commandlet/RepositoryConfig.java
Comment (salimbouch) [reply cli/src/main/java/com/devonfw/tools/ide/commandlet/RepositoryConfig.java:48]: added `trim()` which should remove all leading and ending white spaces and should resolve that problem.

File: cli/src/main/java/com/devonfw/tools/ide/commandlet/RepositoryCommandlet.java
Comment (salimbouch) [reply cli/src/main/java/com/devonfw/tools/ide/commandlet/RepositoryCommandlet.java:132]: Done, removed that if block and will create a new story for importing repositories

File: cli/src/main/java/com/devonfw/tools/ide/commandlet/RepositoryCommandlet.java
Comment (salimbouch) [reply cli/src/main/java/com/devonfw/tools/ide/commandlet/RepositoryCommandlet.java:130]: Done, removed this if block entirely.

File: cli/src/main/java/com/devonfw/tools/ide/property/RepositoryProperty.java
Comment (salimbouch) [reply cli/src/main/java/com/devonfw/tools/ide/property/RepositoryProperty.java:9]: So I can just call `parse` in the RepositoryCommandlet instead? Since Properties have no context attribute how can I access it?

Comment (hohwille) [reply cli/src/main/java/com/devonfw/tools/ide/property/RepositoryProperty.java:10]: Fine for now, but in PR #178 I refactored it to have the IdeContext. I will now also be able to remove the static hacks entirely.

**Comment by hohwille** [comment]
@salimbouch thanks for your reworks. I resolved most of the review comments. 👍 
PR/feature-branch needs to be rebased and conflicts resolved.
Ideally you can even address the final suggestions for `RepositoryProperty`.
As we meanwhile have Coveralls complaining you will need to add a JUnit test covering your new feature.

Once merged, we can further discuss to modernize the API as `ide repository setup` probably does not make so much sense anymore. Will we support other operations than adding (cloning and building) a repository? Either `add` and `remove` would make sense or if we decide to only ever support adding repositories (deletion of a folder can also be done by the user manually) we could even skip the additional keyword (currently it is hardcoded as required what does not make much sense as you cannot use `ide repository ...` with anything else then `setup ...`).

File: cli/src/main/java/com/devonfw/tools/ide/commandlet/RepositoryCommandlet.java
Comment (salimbouch) [reply cli/src/main/java/com/devonfw/tools/ide/commandlet/RepositoryCommandlet.java]: Done

File: cli/src/main/java/com/devonfw/tools/ide/io/FileAccessImpl.java
Comment (hohwille) [inline cli/src/main/java/com/devonfw/tools/ide/io/FileAccessImpl.java:656]: https://github.com/devonfw/IDEasy/blob/main/documentation/coding-conventions.asciidoc#catching-and-handling-exceptions

File: cli/src/main/java/com/devonfw/tools/ide/common/SystemPath.java
Comment (hohwille) [reply cli/src/main/java/com/devonfw/tools/ide/common/SystemPath.java:174]: The "fix" is still incorrect. It will add a path entry and cause duplicates.
Assuming you run the same method twice:
```
systemPath.setPath("mvn", "/d/projects/foo/software/mvn");
systemPath.setPath("mvn","/d/projects/bar/software/mvn");
```
This will cause the `PATH` variable to contain both `/d/projects/foo/software/mvn` and `/d/projects/bar/software/mvn`.
You think this scenario sounds unrealistic?
This is exactly what will happen if I call `ide` in project `foo` and then switch to project `bar` and call `ide` there.
Do we have a bug ticket for this problem itself?

File: cli/src/main/java/com/devonfw/tools/ide/commandlet/RepositoryCommandlet.java
Comment (hohwille) [reply cli/src/main/java/com/devonfw/tools/ide/commandlet/RepositoryCommandlet.java]: Thanks. We still need to implement auto-completion for `RepositoryProperty`. Feel free to do so by overriding the `completeValue` method or to create a new story issue for that so we dont forget this.

File: cli/src/main/java/com/devonfw/tools/ide/commandlet/RepositoryCommandlet.java
Comment (hohwille) [inline cli/src/main/java/com/devonfw/tools/ide/commandlet/RepositoryCommandlet.java:109]: See https://github.com/devonfw/IDEasy/pull/155/files#diff-ca40b72302007b2636f83de91be5b8e9616aab3ba14e6bff7aa7f8d9705d4fb9R57-R60

Comment (salimbouch) [reply cli/src/main/java/com/devonfw/tools/ide/commandlet/RepositoryCommandlet.java:109]: Discussed with @jan-vcapgemini, I'll wait for his PR to be merged then change this block accordingly.

File: cli/src/main/java/com/devonfw/tools/ide/common/SystemPath.java
Comment (salimbouch) [reply cli/src/main/java/com/devonfw/tools/ide/common/SystemPath.java:174]: I just saw your comment right now and couldn't address it in time, good that @MustaphaOuchen took care of it 👍

File: cli/src/main/java/com/devonfw/tools/ide/commandlet/RepositoryCommandlet.java
Comment (salimbouch) [reply cli/src/main/java/com/devonfw/tools/ide/commandlet/RepositoryCommandlet.java:109]: I also noticed the property "build_path" isn't taken into consideration in my implementation. In order to make it work we'd have to change the directory where the process is executed which is not doable since we just call the run() method of the build tool. My suggestion would be we overload the method run() in ToolCommandlet.

File: cli/src/main/java/com/devonfw/tools/ide/commandlet/RepositoryCommandlet.java
Comment (hohwille) [inline cli/src/main/java/com/devonfw/tools/ide/commandlet/RepositoryCommandlet.java:36]: FYI: In devonfw-ide we used the term `setup` as generic keyword for "install":
* `devon mvn setup` installs maven (and Java)
* `devon project setup` installs all configured projects
* We have some logic in the setup process that runs `setup` on all the according commandlets.

However, omitting this keyword would result in `ide repository` what is also rather confusing or not self-explanatory.
We could have the freedom now to change this keyword to anything more suitable.
But I currently do not see a much better name: `ide repository install` might also be kind of confusing.
Potentially we could simply use `add` and then later also have the ability to offer `remove` as well.
Anyway, I will just set this comment to resolved as this seems to be fine and if we come to some better idea, we can still add another PR after merge.

File: cli/src/main/java/com/devonfw/tools/ide/property/RepositoryProperty.java
Comment (hohwille) [inline cli/src/main/java/com/devonfw/tools/ide/property/RepositoryProperty.java:23]:
```suggestion
   */
  public RepositoryProperty(String name, boolean required, String alias) {

    super(name, required, alias, true);
  }
```

File: cli/src/main/java/com/devonfw/tools/ide/property/RepositoryProperty.java
Comment (hohwille) [inline cli/src/main/java/com/devonfw/tools/ide/property/RepositoryProperty.java:37]:
```suggestion
   * @param validator the {@link Consumer} used to {@link #validate() validate} the {@link #getValue() value}.
   */
  public RepositoryProperty(String name, boolean required, String alias, Consumer<Path> validator) {

    super(name, required, alias, true, validator);
  }
```

File: cli/src/main/java/com/devonfw/tools/ide/commandlet/RepositoryCommandlet.java
Comment (hohwille) [inline cli/src/main/java/com/devonfw/tools/ide/commandlet/RepositoryCommandlet.java:37]:
```suggestion
    this.repository = add(new RepositoryProperty("", false, "repository"));
```

File: cli/src/main/java/com/devonfw/tools/ide/commandlet/RepositoryCommandlet.java
Comment (hohwille) [reply cli/src/main/java/com/devonfw/tools/ide/commandlet/RepositoryCommandlet.java:109]: I still think it is better to explicitly pass the branch as separate argument to `GitContext` instead of concatenating url and branch, pass it, and then parse that combined string back into URL and branch. Feels more like a hack than a solution.
But surely it is not wrong and will work this way...

**Review by hohwille — APPROVED** [review]
@salimbouch thanks for your rework and maintaining the feature-branch with all the conflict resolutions.
This now seems fine to me and can be merged.
I left some comments. If you have some time to accept the proposals and maybe even improve on the last suggestion this would be great. I would merge this on Thursday. If you don't address the last point till then, I can merge it and we can still improve this in a separate PR after merge.

File: cli/src/main/java/com/devonfw/tools/ide/commandlet/RepositoryCommandlet.java
Comment (salimbouch) [reply cli/src/main/java/com/devonfw/tools/ide/commandlet/RepositoryCommandlet.java]: Done. also solved bug of `git clone` not working, basically changed `git clone URL#BRANCH` to `git clone URL` then git `checkout BRANCH`. GitUrl which was responsible for parsing the combined string back into URL and branch is now kind of useless, imho we can just delete it.

Comment (hohwille) [reply cli/src/main/java/com/devonfw/tools/ide/commandlet/RepositoryCommandlet.java]: Thanks @salimbouch: You brought this to perfection now 👍

**Pull request merged**
